File: go/leet_code/1249.go
// Package leet_code provides solutions for LeetCode problems.
//
// This file provides a solution for the problem of removing the minimum number of parentheses to make a string valid.
// A string is valid if it is empty, or if it can be built by appending "(", ")", and valid strings.
// A string is balanced if it has as many "(" as ")".
package leet_code

// minRemoveToMakeValid removes the minimum number of parentheses to make a string valid.
//
// It accepts one parameter:
// - s: a string that only contains the characters '(', ')', and alphanumeric characters.
//
// The function returns a string that is a valid parentheses string and is a subsequence of s.
//
// The function uses a stack to keep track of the indices of the '(' characters.
// It iterates over the string, and for each character, if it is '(' it pushes its index to the stack, if it is ')' and the stack is empty, it adds its index to the remove slice, otherwise, it pops an index from the stack.
// After the iteration, it adds the remaining indices in the stack to the remove slice.
// It then constructs the result string by skipping the characters at the indices in the remove slice.
//
// Time complexity analysis:
// - Best-case: O(n), when the string is already valid.
// - Worst-case: O(n), when all characters are '(' or ')'.
// - Average-case: O(n), as we always have to iterate over each character in the string.
func minRemoveToMakeValid(s string) string {
	var stack []int
	var remove []int
	for i, v := range s {
		if v == '(' {
			stack = append(stack, i)
		} else if v == ')' {
			if len(stack) == 0 {
				remove = append(remove, i)
			} else {
				stack = stack[:len(stack)-1]
			}
		}
	}
	remove = append(remove, stack...)
	var res []byte
	for i, v := range s {
		if len(remove) > 0 && i == remove[0] {
			remove = remove[1:]
			continue
		}
		res = append(res, byte(v))
	}
	return string(res)
}

File: java/leet_code/1249.java
import java.util.HashSet;
import java.util.Set;
import java.util.Stack;

/**
 * This class provides a solution for the problem of removing the minimum number of parentheses to
 * make a string valid.
 * A string is valid if it has balanced parentheses.
 * The problem is solved by using a stack to keep track of the indices of the parentheses in the
 * string.
 * If a closing parenthesis is encountered and the stack is empty or the top of the stack is another
 * closing parenthesis, the index of the closing parenthesis is added to a set of indices to be removed.
 * If a closing parenthesis is encountered and the top of the stack is an opening parenthesis,
 * the top of the stack is removed.
 * If an opening parenthesis is encountered, its index is pushed onto the stack.
 * After iterating over the string, any indices remaining in the stack are added to the set of
 * indices to be removed.
 * Finally, a new string is built by appending the characters at the indices not in the set.
 */
class Solution {

    /**
     * Removes the minimum number of parentheses to make a string valid.
     *
     * @param s The input string. It is a string of characters, which can include opening and
     *          closing parentheses.
     *
     * @return The valid string. It is a string of characters, which can include opening and
     *         closing parentheses.
     */
    public String minRemoveToMakeValid(final String s) {
        // Initialize a stack to keep track of the indices of the parentheses in the string
        final Stack<Integer> stack = new Stack<>();

        // Initialize a set to keep track of the indices to be removed
        final Set<Integer> remove = new HashSet<>();

        // Iterate over the characters in the string
        for (int i = 0; i < s.length(); i++) {
            // If an opening parenthesis is encountered, push its index onto the stack
            if ('(' == s.charAt(i)) {
                stack.add(i);
            } else if (')' == s.charAt(i)) {
                // If a closing parenthesis is encountered and the stack is empty or the top of the
                // stack is another closing parenthesis,
                // add the index of the closing parenthesis to the set of indices to be removed
                if (stack.isEmpty() || ')' == s.charAt(stack.peek())) {
                    remove.add(i);
                } else {
                    // If a closing parenthesis is encountered and the top of the stack is an
                    // opening parenthesis,
                    // remove the top of the stack
                    stack.pop();
                }
            }
        }

        // Add any indices remaining in the stack to the set of indices to be removed
        while (!stack.isEmpty()) {
            remove.add(stack.pop());
        }

        // Initialize a string builder
        final StringBuilder sb = new StringBuilder();

        // Append the characters at the indices not in the set to the string builder
        for (int i = 0; i < s.length(); i++) {
            if (!remove.contains(i)) {
                sb.append(s.charAt(i));
            }
        }

        // Return the string builder as a string
        return sb.toString();
    }
}
